src/lib.ml

```diff
@@ -877,8 +955,122 @@ and resolve_user_deps db deps ~allow_private_deps ~pps ~stack =
+      begin
+        match Lib_name.Map.find !vlib_dfs_status lib.name with
+        | Some (Some ()) ->
+          Error (Error (Default_implementation_cycle
+                          { cycle = (lib.info::stack)
+                          }))
+        | Some None -> Ok ()
+        | None ->
+          begin
+            (* Exploring node. *)
+            vlib_dfs_status :=
+              Lib_name.Map.add !vlib_dfs_status lib.name (Some ());
+            (* Visit direct dependencies *)
+            lib.requires >>= fun deps ->
+            (List.filter ~f:(fun x ->
+               match avoid_direct_parent x lib with
+               | Ok x -> x
+               | Error _ -> false) deps)
+            |> visit ~stack:(lib.info::stack) ancestor_vlib
+            >>= fun () ->
+            (* If the library is an implementation of some virtual library that
+               overrides default, add a link in the graph. *)
+            Option.map lib.implements
+              ~f:(fun vlib ->
+                vlib >>= fun vlib ->
+                impl_different_from_vlib_default vlib lib >>= function res ->
+                match (res, ancestor_vlib) with
+                | true, None ->
+                  (* Recursion: no ancestor, vlib is explored *)
+                  visit ~stack:(lib.info::stack) None [vlib]
+                | true, Some ancestor ->
+                  vlib_default_parent := Lib_name.Map.update
+                                           !vlib_default_parent
+                                           lib.name
+                                           ~f:(merge ancestor.name);
+                  visit ~stack:(lib.info::stack) None [vlib]
+                | false, _ ->
+                  (* If lib is the default implementation,
+                     we'll manage it when handling virtual lib. *)
+                  Ok ())
+            |> Option.value ~default:(Ok ()) >>= fun () ->
+            (* If the library has an implementation according to variants. *)
+            get_default_implementation lib >>= fun default_implementation ->
+            Option.map default_implementation ~f:(fun (loc,name) ->
+              name_to_lib name loc >>= fun default_impl ->
+              visit ~stack:(lib.info::stack) (Some lib) [default_impl])
+            |> Option.value ~default:(Ok ()) >>= fun () ->
+            (* If the library is a virtual library with a default
+               implementation. *)
+            vlib_dfs_status := Lib_name.Map.add !vlib_dfs_status lib.name None;
+            visit ~stack ancestor_vlib next
+          end
+      end
+  in
+  (* For each virtual library we know which vlibs will be implemented when
+     enabling its default implementation. *)
+  visit ~stack:[] None libraries >>= fun () ->
+  Ok (List.filter_map ~f:library_is_default libraries)
+
+and closure_with_overlap_checks db ts ~stack:orig_stack ~linking ~variants =
+  let name_to_lib name loc =
+    resolve_dep (Option.value_exn db) name
```

Comment:
This can't be right I think. `db` is None in some cases. Why can't it be `None` here?

Reply:
I agree this is weird. I assumed the database was accessible when virtual libraries resolution was needed. We just need to be sure that `variants` is empty when `db` is `None`. It's correct for every case, except in `lib.ml:  resolve_user_written_deps_for_exe`, where `db` is `(Option.some_if (not allow_overlaps) t)`. I don't really know about that `allow_overlap` feature: when is it set to true ?